velox/exec/Task.cpp

```diff
@@ -1785,6 +1838,22 @@ ContinueFuture Task::stateChangeFuture(uint64_t maxWaitMicros) {
   return std::move(future);
 }
 
+ContinueFuture Task::taskCompletionFuture(uint64_t maxWaitMicros) {
```

Comment:
@gggrace14 
Thanks for looking into it!
I don't think we can just use one even at the moment.
This can break tests.
There are also use cases of the old logic in Meta internal code.
If, in the future, we don't need the finish - we can remove it.
Right now I prefer to proceed safely and differentiate the two events. :)

Reply:
@spershin , just to clarify, do I understand it correctly that we introduce taskCompletionPromises_ but it's not in execution path and is only for tests? I'm okay if you think it works.

Another part is about synchronization and for me to understand. Is it a possible case that the worker is blocked by two getTaskStatus requests and there're two promises in the stateChangePromises_? Then we receive noMoreSplitsForGroup() for only one split group, noMoreSplitsForGroup() calls stateChangeNotifier.notify() and set both of the two promises of stateChangePromises_. Thus the worker responds to two getTaskStatus requests and then the coordinator subsequently sends two more split groups, even when the worker is still processing the remaining one split group.

I guess the above scenario is not precisely correct, but we're okay with it. It's like finishing split groups more aggresively.

Reply:
@gggrace14 
We should sync offline. It is tedious to exchange messages here.